### ISSUE TITLE
L034: Columns should not be re-ordered / misidentified Aggregation function

### ISSUE THREAD

**Issue by kevin868** [sqlfluff/sqlfluff#3418]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

The columns are swapped when formatting a small query, seemingly to push Aggregations together below regular columns. This is problematic, when using functions like `DATE_TRUNC()` or LEFT(string_field, 5) which are scalar row-ops. Meanwhile, the GROUP BY numbers do not change, so the whole query meaning is messed up.

Input:

```sql
SELECT
    DATE_TRUNC('DAY', end_time) AS time_day,
   b_field,
    c_field,
    COUNT(*) AS e_field,
    SUM(g_field) AS sum_g_field
 FROM table_name
 where end_time >= '2022-05-01'
GROUP BY 1,2,3
```

I looked around on Slack. Apologies if this is not a bug but a simple config issue. Just seemed very surprising and breaking behavior.

### Expected Behaviour

Here, the only issue is `b_field` whitespace should match all the others. We expect that as the only fix.

```sql
SELECT
    DATE_TRUNC('DAY', end_time) AS time_day,
    b_field,
    c_field,
    COUNT(*) AS e_field,
    SUM(g_field) AS sum_g_field
 FROM table_name
 where end_time >= '2022-05-01'
GROUP BY 1,2,3
```

### Observed Behaviour

DATE_TRUNC... gets pushed to column 3 (`b_field` and `c_field` move up to columns 1, 2)

```sql
SELECT
    b_field,
    c_field,
    DATE_TRUNC('DAY', end_time) AS time_day,
    COUNT(*) AS e_field,
    SUM(g_field) AS sum_g_field
FROM table_name
WHERE end_time >= '2022-05-01'
GROUP BY 1, 2, 3
```

### How to reproduce

`sqlfluff fix --force file.sql`

### Dialect

ansi, snowflake

### Version

0.13.2

### Configuration

```
[sqlfluff]
dialect = ansi

[sqlfluff:rules]
max_line_length = 120
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by barrywhart** [comment]
To fix this, we can probably leverage parts of `L054`, which checks for use of numeric column references.

**Comment by kevin868** [comment]
Thanks for the quick turnaround!  I think L034 solves a legitimate issue, but I worry - does it still get applied when the SELECT happens for an INSERT or MERGE statement?
Thus, even though your fix stops it for numeric column references.. I would still think the column reordering is problematic for INSERT-type statements.  These expect a specific column order, that match the table definition.

Your PR looks fine, but I don't understand the codebase well enough to comment :)

Please let me know if you think that proper L034 operation is safe from the above issue. (or possibly open a new one to address that) Thanks again to both of you, @tunetheweb, @barrywhart.

**Comment by barrywhart** [comment]
@kevin868: There's an [existing test case](https://github.com/sqlfluff/sqlfluff/blob/main/test/fixtures/rules/std_rule_cases/L034.yml#L105) for ignoring INSERT statements, but not for merge. Please create a separate issue for ignoring merge statements.

**Comment by tunetheweb** [comment]
@barrywhart this seems a simple thing we should fix.

Could we not add it to #3423 by changing this line:

```
            if len(context.parent_stack) >= 3 and context.parent_stack[-3].is_type(
                "create_table_statement"
            ):
                return None
```

to:

```py
            if len(context.parent_stack) >= 3 and context.parent_stack[-3].is_type(
                "merge_statement", "create_table_statement"
            ):
                return None
```

Here's an example sql, that's currently broken, but should remain untouched with above fix:

```sql
MERGE INTO t
USING
(
    SELECT
        DATE_TRUNC('DAY', end_time) AS time_day,
        b
    FROM u
) AS u ON (a = b)
WHEN MATCHED THEN
UPDATE SET a = b
WHEN NOT MATCHED THEN
INSERT (b) VALUES (c);
```

**Comment by barrywhart** [comment]
I'll add it. In general, I try to avoid adding scope to these smaller issues once I've started writing code, because it makes it harder to get things finished review and merge.

**Comment by barrywhart** [comment]
Ok, added support for ignoring `MERGE` statements.
